### ISSUE TITLE
Crashes on Android 11 (Gcam >=7.5)

### ISSUE THREAD

**Issue by losuler** [lukaspieper/Gcam-Services-Provider#6]
Pixel 4a, RattlesnakeOS, Android 11 (`RP1A.201005.006.2020.10.06.13`)

Using the .apk files from https://www.apkmirror.com/apk/google-inc/camera, the `7.4.x` apk's don't crash but any of the `7.5.x` .apk files do (`7.5.105.323030203` and `7.5.107.326355469`). I tried both the [latest release](https://github.com/lukaspieper/Gcam-Services-Provider/releases/tag/v1.0) of Gcam-Services-Provider and built from source with the same result.

This is what logcat outputs:
```
10-07 14:42:02.448  1637  1867 I ActivityManager: Start proc 10405:com.google.android.GoogleCamera/u0a190 for pre-top-activity {com.google.android.GoogleCamera/com.android.camera.CameraLauncher}
10-07 14:42:02.452 10405 10405 E id.GoogleCamer: Not starting debugger since process cannot load the jdwp agent.
10-07 14:42:02.469 10405 10405 I id.GoogleCamer: The ClassLoaderContext is a special shared library.
10-07 14:42:02.470 10405 10405 D nativeloader: classloader namespace configured for unbundled vendor apk. library_path=/data/app/~~gvlUQxdNGgx3EPXeRZgdJA==/com.google.android.GoogleCamera-_SUWbO4PLgFR_eiFLXQChQ==/lib/arm64:/data/app/~~gvlUQxdNGgx3EPXeRZgdJA==/com.google.android.GoogleCamera-_SUWbO4PLgFR_eiFLXQChQ==/base.apk!/lib/arm64-v8a:/vendor/lib64
10-07 14:42:02.481 10405 10405 D NetworkSecurityConfig: No Network Security Config specified, using platform default
10-07 14:42:02.481 10405 10405 D NetworkSecurityConfig: No Network Security Config specified, using platform default
10-07 14:42:02.490 10405 10405 W CAM_SearchIndexProv: Called onCreate
10-07 14:42:02.497 10405 10405 D AndroidRuntime: Shutting down VM
10-07 14:42:02.497 10405 10405 E AndroidRuntime: FATAL EXCEPTION: main
10-07 14:42:02.497 10405 10405 E AndroidRuntime: Process: com.google.android.GoogleCamera, PID: 10405
10-07 14:42:02.497 10405 10405 E AndroidRuntime: java.lang.RuntimeException: Unable to create application com.google.android.apps.camera.legacy.app.app.CameraApp: java.lang.SecurityException: Failed to find provider com.google.android.gsf.gservices for user 0; expected to find a valid ContentProvider for this authority
10-07 14:42:02.497 10405 10405 E AndroidRuntime: 	at android.app.ActivityThread.handleBindApplication(ActivityThread.java:6717)
10-07 14:42:02.497 10405 10405 E AndroidRuntime: 	at android.app.ActivityThread.access$1300(ActivityThread.java:237)
10-07 14:42:02.497 10405 10405 E AndroidRuntime: 	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1913)
10-07 14:42:02.497 10405 10405 E AndroidRuntime: 	at android.os.Handler.dispatchMessage(Handler.java:106)
10-07 14:42:02.497 10405 10405 E AndroidRuntime: 	at android.os.Looper.loop(Looper.java:223)
10-07 14:42:02.497 10405 10405 E AndroidRuntime: 	at android.app.ActivityThread.main(ActivityThread.java:7656)
10-07 14:42:02.497 10405 10405 E AndroidRuntime: 	at java.lang.reflect.Method.invoke(Native Method)
10-07 14:42:02.497 10405 10405 E AndroidRuntime: 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
10-07 14:42:02.497 10405 10405 E AndroidRuntime: 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947)
10-07 14:42:02.497 10405 10405 E AndroidRuntime: Caused by: java.lang.SecurityException: Failed to find provider com.google.android.gsf.gservices for user 0; expected to find a valid ContentProvider for this authority
10-07 14:42:02.497 10405 10405 E AndroidRuntime: 	at android.os.Parcel.createExceptionOrNull(Parcel.java:2373)
10-07 14:42:02.497 10405 10405 E AndroidRuntime: 	at android.os.Parcel.createException(Parcel.java:2357)
10-07 14:42:02.497 10405 10405 E AndroidRuntime: 	at android.os.Parcel.readException(Parcel.java:2340)
10-07 14:42:02.497 10405 10405 E AndroidRuntime: 	at android.os.Parcel.readException(Parcel.java:2282)
10-07 14:42:02.497 10405 10405 E AndroidRuntime: 	at android.content.IContentService$Stub$Proxy.registerContentObserver(IContentService.java:1229)
10-07 14:42:02.497 10405 10405 E AndroidRuntime: 	at android.content.ContentResolver.registerContentObserver(ContentResolver.java:2627)
10-07 14:42:02.497 10405 10405 E AndroidRuntime: 	at android.content.ContentResolver.registerContentObserver(ContentResolver.java:2615)
10-07 14:42:02.497 10405 10405 E AndroidRuntime: 	at lpr.b(PG:14)
10-07 14:42:02.497 10405 10405 E AndroidRuntime: 	at lpr.a(PG:69)
10-07 14:42:02.497 10405 10405 E AndroidRuntime: 	at lpr.a(PG:33)
10-07 14:42:02.497 10405 10405 E AndroidRuntime: 	at com.google.android.apps.camera.legacy.app.app.CameraApp.onCreate(PG:115)
10-07 14:42:02.497 10405 10405 E AndroidRuntime: 	at android.app.Instrumentation.callApplicationOnCreate(Instrumentation.java:1192)
10-07 14:42:02.497 10405 10405 E AndroidRuntime: 	at android.app.ActivityThread.handleBindApplication(ActivityThread.java:6712)
10-07 14:42:02.497 10405 10405 E AndroidRuntime: 	... 8 more
10-07 14:42:02.497 10405 10405 E AndroidRuntime: Caused by: android.os.RemoteException: Remote stack trace:
10-07 14:42:02.497 10405 10405 E AndroidRuntime: 	at com.android.server.content.ContentService.registerContentObserver(ContentService.java:355)
10-07 14:42:02.497 10405 10405 E AndroidRuntime: 	at android.content.IContentService$Stub.onTransact(IContentService.java:482)
10-07 14:42:02.497 10405 10405 E AndroidRuntime: 	at android.os.Binder.execTransactInternal(Binder.java:1154)
10-07 14:42:02.497 10405 10405 E AndroidRuntime: 	at android.os.Binder.execTransact(Binder.java:1123)
10-07 14:42:02.497 10405 10405 E AndroidRuntime: 
10-07 14:42:02.499  1637 10427 I DropBoxManagerService: add tag=data_app_crash isTagEnabled=true flags=0x2
10-07 14:42:02.499  1637  2395 W ActivityTaskManager:   Force finishing activity com.google.android.GoogleCamera/com.android.camera.CameraLauncher
10-07 14:42:02.519  1637  1854 D EventSequenceValidator: onActivityLaunchCancelled during UNKNOWN.
10-07 14:42:02.519  1637  1854 D EventSequenceValidator: java.lang.Throwable: EventSequenceValidator#getStackTrace
10-07 14:42:02.519  1637  1854 D EventSequenceValidator: 	at com.google.android.startop.iorap.EventSequenceValidator.logWarningWithStackTrace(EventSequenceValidator.java:260)
10-07 14:42:02.519  1637  1854 D EventSequenceValidator: 	at com.google.android.startop.iorap.EventSequenceValidator.onActivityLaunchCancelled(EventSequenceValidator.java:166)
10-07 14:42:02.519  1637  1854 D EventSequenceValidator: 	at com.android.server.wm.LaunchObserverRegistryImpl.handleOnActivityLaunchCancelled(LaunchObserverRegistryImpl.java:166)
10-07 14:42:02.519  1637  1854 D EventSequenceValidator: 	at com.android.server.wm.LaunchObserverRegistryImpl.lambda$lAGPwfsXJvBWsyG2rbEfo3sTv34(Unknown Source:0)
10-07 14:42:02.519  1637  1854 D EventSequenceValidator: 	at com.android.server.wm.-$$Lambda$LaunchObserverRegistryImpl$lAGPwfsXJvBWsyG2rbEfo3sTv34.accept(Unknown Source:4)
10-07 14:42:02.519  1637  1854 D EventSequenceValidator: 	at com.android.internal.util.function.pooled.PooledLambdaImpl.doInvoke(PooledLambdaImpl.java:278)
10-07 14:42:02.519  1637  1854 D EventSequenceValidator: 	at com.android.internal.util.function.pooled.PooledLambdaImpl.invoke(PooledLambdaImpl.java:201)
10-07 14:42:02.519  1637  1854 D EventSequenceValidator: 	at com.android.internal.util.function.pooled.OmniFunction.run(OmniFunction.java:97)
10-07 14:42:02.519  1637  1854 D EventSequenceValidator: 	at android.os.Handler.handleCallback(Handler.java:938)
10-07 14:42:02.519  1637  1854 D EventSequenceValidator: 	at android.os.Handler.dispatchMessage(Handler.java:99)
10-07 14:42:02.519  1637  1854 D EventSequenceValidator: 	at android.os.Looper.loop(Looper.java:223)
10-07 14:42:02.519  1637  1854 D EventSequenceValidator: 	at android.os.HandlerThread.run(HandlerThread.java:67)
10-07 14:42:02.519  1637  1854 D EventSequenceValidator: 	at com.android.server.ServiceThread.run(ServiceThread.java:44)
10-07 14:42:02.519  1637  1854 D EventSequenceValidator: dec AccIntentStartedEvents to 1
10-07 14:42:02.622  3246  3246 W Choreographer: Frame time is 0.002495 ms in the future!  Check that graphics HAL is generating vsync timestamps using the correct timebase.
10-07 14:42:02.656  3246  3246 W Choreographer: Frame time is 0.007022 ms in the future!  Check that graphics HAL is generating vsync timestamps using the correct timebase.
10-07 14:42:02.689  3246  3246 W Choreographer: Frame time is 0.016748 ms in the future!  Check that graphics HAL is generating vsync timestamps using the correct timebase.
10-07 14:42:02.722  3246  3246 W Choreographer: Frame time is 0.037457 ms in the future!  Check that graphics HAL is generating vsync timestamps using the correct timebase.
10-07 14:42:02.722  5057  5057 W Choreographer: Frame time is 0.024332 ms in the future!  Check that graphics HAL is generating vsync timestamps using the correct timebase.
10-07 14:42:02.789  3246  3246 W Choreographer: Frame time is 0.011106 ms in the future!  Check that graphics HAL is generating vsync timestamps using the correct timebase.
10-07 14:42:02.836 15260 15305 I Heartbeat: Sending Heartbeat
10-07 14:42:02.853 15260 15305 I Heartbeat: Received Heartbeat with 16ms latency
10-07 14:42:02.905  3246  3246 W Choreographer: Frame time is 0.011732 ms in the future!  Check that graphics HAL is generating vsync timestamps using the correct timebase.
10-07 14:42:02.956   623   623 E Layer   : [Surface(name=Task=1635)/@0x72ffb49 - animation-leash#0] No local sync point found
10-07 14:42:02.956   623   623 E Layer   : [Surface(name=Task=1)/@0xcda94d - animation-leash#0] No local sync point found
10-07 14:42:03.000  1637  1854 W ActivityTaskManager: Activity top resumed state loss timeout for ActivityRecord{1241f23 u0 com.google.android.GoogleCamera/com.android.camera.CameraLauncher t1635 f}}
10-07 14:42:03.014  1637  1854 W ActivityTaskManager: Activity pause timeout for ActivityRecord{1241f23 u0 com.google.android.GoogleCamera/com.android.camera.CameraLauncher t1635 f}}
```

**Comment by lukaspieper** [comment]
Thanks for reporting. 

If I'm not mistaken this is exactly the same exception that occurs when you run Gcam 7.4.x without Gcam-Services-Provider. Therefor I would assume that it is not a bug in the implementation. Rather it could be possible that Gcam cannot communicate with my stub implementation any longer.

Sadly I can't look into this at the moment because I just don't have enough time for it. Though the good news is that Gcam 7.4.x runs on Android 11.

**Comment by ProjectMoon** [comment]
I ran into this as well, and for the fun of it I installed microG services core as a user app on a ROM with no signature spoofing. All permissions are disabled. With microG installed, Google Camera 7.5 runs just fine. So there must be some new APIs that require implementation in this project.

**Comment by lukaspieper** [comment]
@ProjectMoon  Thanks for bringing this to my attention. I pushed a fix and created a new release.

**Comment by lukaspieper** [comment]
Reopening this based on the discussion (#16).

**Comment by lukaspieper** [comment]
For all of you building the app from source, if you set `--force-queryable` as install flag in Android Studio, it will work again with version 7.5. 

![image](https://user-images.githubusercontent.com/30661176/101825958-a2c0a280-3b2e-11eb-914f-ad6a55b95628.png)

**Comment by innir** [comment]
> Edit: Has anyone who is running into this issue has tried re-installing the services-provider?

Yes, I did that (but without your proposed changes) and it didn't work.

**Comment by innir** [comment]
For me something along these lines worked:

```
adb push app-debug.apk /data/local/tmp
adb shell pm install --force-queryable /data/local/tmp/app-debug.apk
```
Compiling the app yourself is not necessary (but recommended anyways).

**Comment by maogr235** [comment]
> For all of you building the app from source, if you set `--force-queryable` as install flag in Android Studio, it will work again with version 7.5.
> 
> ![image](https://user-images.githubusercontent.com/30661176/101825958-a2c0a280-3b2e-11eb-914f-ad6a55b95628.png)
> 
> Edit: Has anyone who is running into this issue has tried re-installing the services-provider?

Hi, sorry to bother you. I tried to build the apk with that flag but it didnt help. It took me 2 hours to find where i can write that --force-queryable based on your screenshot so its really possible i am doing something wrong. But i managed to built the apk month ago. It worked until last update. Could you upload here compiled apk with this flag forced? Or this is not how its work?

**Comment by lukaspieper** [comment]
No need to apologize.

Force-queryable is already included in my Android 11 apk (1.1). Apparently, however, the value is no longer honored by the Android OS. Therefore, I can not provide a new APK, there would be no differences. The installation flags only affect the installation process. 

But the commands from @innir should work with a prebuild apk. If you don't want to use a apk from releases you can also get apks from GitHub Actions, e. g. [here](https://github.com/lukaspieper/Gcam-Services-Provider/actions/runs/379245816) is the latest for now.

**Comment by maogr235** [comment]
your comment made me realize i should connect my phone to the pc and install apk using android studio, not sideloading it like i was until now. Thank you very much, now it works perfectly like it was before.

**Comment by beerisgood** [comment]
So their is no way without Android Studio or using ADB ?

**Comment by achterstag** [comment]
`adb shell pm install --force-queryable /data/local/tmp/app-debug.apk`
did the trick for me. Thank you!

**Comment by lukaspieper** [comment]
> So their is no way without Android Studio or using ADB ?

For version 7.5 I don't know any other solution at the moment. Otherwise use version 7.4.

**Comment by Terminator-J** [comment]
> So their is no way without Android Studio or using ADB ?

I mean, you could probably do it from the handset in a root shell in Terminal or Termux, as long as you've copied the apk to internal storage somewhere (for example, /sdcard/Download), then just `pm install --force-queryable /sdcard/Download/app-debug.apk` and be good to go. I know this is necroposting, but it's still something folks are dealing with, and your release still works; just had a user do this to get Arnova82's latest GCam 8.1 port working successfully without any GApps/microG installed on crDroid 7.5 (android 11) on a OnePlus 6T.

**Comment by sorryusernameisalreadytaken** [comment]
> `adb shell pm install --force-queryable /data/local/tmp/app-debug.apk`
> did the trick for me. Thank you!

When I try this, I just got:
```
adb shell pm install --force-queryable /data/local/tmp/app-debug.apk
>>Failure [INSTALL_FAILED_VERSION_DOWNGRADE]
```
and
`pm list packages | grep lukas`
show 0 results...

**Comment by lukaspieper** [comment]
Calyx released an app that adds support for the preview button. Their app also contains code that makes `--force-queryable` in my app redundant. Please try my new APK from [here](https://github.com/lukaspieper/Gcam-Services-Provider/actions/runs/1295325750). To confirm that the installation parameter is no longer needed, you first need to uninstall Gcam-Services-Provider manually. After that the "basic" apk should work without any tricks.

[Here](https://github.com/lukaspieper/Gcam-Services-Provider/issues/11#issuecomment-932261437) can you read more about the upcoming "photos" apk.

**Comment by gryban1** [comment]
Works! (Pixel 5, GrapheneOS, GCam 8.2.400)

**Comment by lukaspieper** [comment]
[Fixed with version 1.3](https://github.com/lukaspieper/Gcam-Services-Provider/releases/tag/v1.3)
